### ISSUE TITLE
Too many subgraphs. Teach me HOW to compile an own made model into an xmodel with only one subgraph. 

### ISSUE THREAD

**Issue by Hikaru-Furuta** [Xilinx/Vitis-AI#336]
I'm trying to compile an own customized xmodel for ZCU104  in a conda env, vitis-ai-pytorch, using Vitis-ai docker.
I know we need a quantized xmodel and arch.json for ZCU104 to generate an xmodel for inference on ZCU104, but they were converted into an xmodel, which has 52 root subgraphs.
I just want them to be only one subgraph.
And they seem to be separated between Conv2d and LeakyRelu which are supported in DPUCZDX8G(shown by Table24 on page 90  in the Vitis ai User guide : UG1414 v1.3 December 17, 2020).
Will the methods described on Page 86~88 in the User guide work to solve this problem?
If there's any other way out to compile the model into an xmodel with only one root subgraph, please teach me.

**Comment by qianglin-xlnx** [comment]
Hi @Hikaru-Furuta 
Which pytorch model do you use?

**Comment by Hikaru-Furuta** [comment]
@qianglin-xlnx 
thanks for replying
The one I've created by myself.
after extracting Leakyrelu from the network, the number of subgraphs became 4.
I think Leakyrelus cannot be converted by this compiler right?

**Comment by Fabioni** [comment]
Hey @Hikaru-Furuta and @qianglin-xlnx,

how did you solve the problem? I have the same problem when compiling my own pytorch model. I have 152 subgraphs. Each one look like this:

```
in:  ["{name: 'Darknet__Darknet_Sequential_LeakyReLU_activation__input_298_fix', shape: [1, 32, 32, 256], type: 'xint8', attrs: {'ddr_addr': 0, 'fix_point': 4, 'location': 1, 'round_mode': 'DPU_ROUND', 'bit_width': 8, 'if_signed': True, 'reg_id': 1}}[1, 32, 32, 256]"]
out:  ["{name: 'Darknet__Darknet_Sequential_Conv2d_Conv2d__input_302_fix', shape: [1, 16, 16, 512], type: 'xint8', attrs: {'ddr_addr': 262144, 'fix_point': 4, 'location': 1, 'round_mode': 'DPU_ROUND', 'bit_width': 8, 'if_signed': True, 'reg_id': 1}}[1, 16, 16, 512]"]
```

I only want to have one subgraph.

**Comment by Hikaru-Furuta** [comment]
@Fabioni  
I couldn't solve, but think these problems are caused by unsupported Layers like leaky relu.
In order to make it in one graph, we may need to use only supported Layers(padding, concat, activation functions, etc.)

**Comment by Fabioni** [comment]
Hey, I also found the problem today. In Vitis ai user guide, they say LeakyRelu is supported but obviously this was the problem. Exchanging LeakyRelu to Relu solved the problem with the subgraphs. Upsampeling was also a problem but disappeared somehow, I really don't know why :D
Now I have 5 subgraphs and one on the DPU. First I thought thats a problem but actually that is how it should be. One is the input, three are the output and one is the actual subgraph we want.

For the moment I will just go with that (so Relu) but would really be interested in why LeakyRelu doesn't work here although it is mentioned as support in the user guide.

Greetings,
Fabian

**Comment by Fabioni** [comment]
Maybe it is just with pytorch and it works if we go the detour ober caffee or tensorflow.

**Comment by Hikaru-Furuta** [comment]
@Fabioni 
Yes, exactly. I've seen that leaky relu is mentioned as supported function in the user guide too.
As for the outputs, I've got the same problem as well.
I meant it to be in only one output and wanted to make only one subgraph but it separated into three elements when compiled.
anyway, thank you for the  information!

**Comment by Fabioni** [comment]
The output is no problem, my model outputs a tuple of three tensors so that is really what I want. It's just a bit confusing to call this a subgraph.

**Comment by Hikaru-Furuta** [comment]
@Fabioni 
and I only have one more question. What DPU do you use for your DNN?

**Comment by Fabioni** [comment]
I use the pre build image of the zcu102

**Comment by Fabioni** [comment]
I use the pre build image of the zcu102

**Comment by Hikaru-Furuta** [comment]
Thank you

**Comment by BraunPhilipp** [comment]
I highly recommend looking into the xnnc compiler code. All supported ops are clearly stated. Caffe has the best support in my opinion. Otherwise, you will need to fold your batch norm etc. into a conv layer.

* https://github.com/Xilinx/Vitis-AI/blob/dd383346871339e5ee1ffbdec3bfa665210fcfd2/docs/quick-start/install/Vitis%20AI%201.3.2%20April%202021%20Patch.md

**Comment by jackliu-bili** [comment]
Hi,everyone
I have same problem.
I met a AssertionError because "assert len(subgraphs) == 1".
Now, I want to solve the problem but I havn't see a perfect solution.There‘s a possibility that these problems are caused by unsupported Layers. But I don't know which layers are unsupported layers.
So, I pull the image here. Hoping that you can help me.
Thanks very mach.
![1634662389(1)](https://user-images.githubusercontent.com/56680352/137956604-6008285e-3502-48f0-8d71-ea657c699624.jpg)

**Comment by levent9402** [comment]
> 

Hi，I find your model has too many subgraphs, but the demo code only support 1 subgraphs. It is possible that there are unsupported layers in the model.

**Comment by guohaot-Xlnx** [comment]
Hi @Hikaru-Furuta ,
This issue is about to close. If you have any questions, please repost to the issues
